src/verify.c

```diff
@@ -0,0 +1,131 @@
+#include <openssl/bio.h>
+#include <openssl/cms.h>
+#include <openssl/err.h>
+#include <openssl/pem.h>
+#include <openssl/safestack.h>
+#include <openssl/x509.h>
+#include <openssl/x509_vfy.h>
+
+#include "./verify.h"
+
+return_t VF_init() {
+  ERR_load_crypto_strings();
+  OpenSSL_add_all_algorithms();
+  // Since none of these functions return useful error messages, per the
+  // openssl wiki documentation, we're just going to return success status.
+  // If this changes in future, we have the option to start failing
+  return VF_SUCCESS;
+}
+
+// TODO:
+//  - make sure all cleanup happens all the time
+//  - make sure that all return values are correct read
+//  - consider using ERR_get_error (iirc) to get actual error reasons
+//  - make sure we're correctly forcing FORMAT_PEM
+return_t VF_verify(char *pubkey, uint64_t pubkey_l, char *document,
+                   uint64_t document_l, char *pkcs7, uint64_t pkcs7_l,
+                   struct Error **err) {
+  // We want to clear the OpenSSL Error queue so that we know when we're in the
+  // cleanup section, any errors we hit are the result of this invocation
+  ERR_clear_error();
+  return_t rv;
+
+  // First, we need to convert from C-strings into BIOs backed by memory
+  BIO *bio_pubkey = BIO_new_mem_buf(pubkey, pubkey_l);
+  BIO *bio_document = BIO_new_mem_buf(document, document_l);
+  BIO *bio_pkcs7 = BIO_new_mem_buf(pkcs7, pkcs7_l);
+
+  // We don't want the BIO_close method here to free() the memory passed in
+  // BIO_set_close(bio_pubkey, BIO_NOCLOSE);
+  // BIO_set_close(bio_document, BIO_NOCLOSE);
+  // BIO_set_close(bio_pkcs7, BIO_NOCLOSE);
+
+  // The data structures needed for verification
+  PKCS7 *p7;
+  X509_STORE *store = NULL;
+  STACK_OF(X509) *certs = NULL;
+  X509 *cert = NULL;
+
+  // I'd prefer to pass an out parameter, but it seems that the second time
+  // that this function executes, we get a segfault in the underlying library.
+  // I'm not sure how best to debug this, but using the pointer returning
+  // option of the function works well
+  //
+  //   if(!PEM_read_bio_PKCS7(bio_pkcs7, &p7, NULL, NULL)) {
+  //
```

Comment:
The docs say:

> The PEM read functions all take an argument TYPE **x and return a TYPE * pointer. Where TYPE is whatever structure the function uses. If x is NULL then the parameter is ignored. If x is not NULL but *x is NULL then the structure returned will be written to *x. If neither x nor *x is NULL then an attempt is made to reuse the structure at *x (but see BUGS and EXAMPLES sections). Irrespective of the value of x a pointer to the structure is always returned (or NULL if an error occurred).

I suspect that this crash is because `p7` is not initialized to `NULL`.  But using the return value is also fine, so you can probably just remove this comment :)

The lack of NULL initialization could also potentially cause a crash calling `PKCS7_free(..)` on a garbage value

Reply:
Indeed. `p7` should be nulled on initialised independent on how you use it.

Reply:
ahh, that makes sense.  I'll `NULL` it.